src/jit/codegenarmarch.cpp

```diff
@@ -1924,149 +1924,266 @@ void CodeGen::genCodeForCpBlkHelper(GenTreeBlk* cpBlkNode)
+            assert(srcAddr->OperIsLocalAddr());
+            srcLclNum = srcAddr->AsLclVarCommon()->GetLclNum();
+            srcOffset = srcAddr->OperIs(GT_LCL_FLD_ADDR) ? srcAddr->AsLclFld()->gtLclOffs : 0;
         }
-        srcAddr = source;
     }
 
-    unsigned offset = 0;
+    if (node->IsVolatile())
+    {
+        instGen_MemoryBarrier();
+    }
+
+    emitter* emit = GetEmitter();
+    unsigned size = node->GetLayout()->GetSize();
 
-    // Grab the integer temp register to emit the loads and stores.
-    regNumber tmpReg = cpBlkNode->ExtractTempReg(RBM_ALLINT);
+    regNumber tempReg = node->ExtractTempReg(RBM_ALLINT);
 
 #ifdef _TARGET_ARM64_
     if (size >= 2 * REGSIZE_BYTES)
     {
-        regNumber tmp2Reg = cpBlkNode->ExtractTempReg(RBM_ALLINT);
+        regNumber tempReg2 = node->ExtractTempReg(RBM_ALLINT);
 
-        size_t slots = size / (2 * REGSIZE_BYTES);
-
-        while (slots-- > 0)
+        for (unsigned regSize = 2 * REGSIZE_BYTES; size >= regSize;
+             size -= regSize, srcOffset += regSize, dstOffset += regSize)
         {
-            // Load
-            genCodeForLoadPairOffset(tmpReg, tmp2Reg, srcAddr, offset);
-            // Store
-            genCodeForStorePairOffset(tmpReg, tmp2Reg, dstAddr, offset);
-            offset += 2 * REGSIZE_BYTES;
+            if (srcLclNum != BAD_VAR_NUM)
+            {
+                emit->emitIns_R_R_S_S(INS_ldp, EA_8BYTE, EA_8BYTE, tempReg, tempReg2, srcLclNum, srcOffset);
+            }
+            else
+            {
+                emit->emitIns_R_R_R_I(INS_ldp, EA_8BYTE, tempReg, tempReg2, srcAddrBaseReg, srcOffset);
+            }
+
+            if (dstLclNum != BAD_VAR_NUM)
+            {
+                emit->emitIns_S_S_R_R(INS_stp, EA_8BYTE, EA_8BYTE, tempReg, tempReg2, dstLclNum, dstOffset);
+            }
+            else
+            {
+                emit->emitIns_R_R_R_I(INS_stp, EA_8BYTE, tempReg, tempReg2, dstAddrBaseReg, dstOffset);
+            }
         }
     }
+#endif
 
-    // Fill the remainder (15 bytes or less) if there's one.
-    if ((size & 0xf) != 0)
+    for (unsigned regSize = REGSIZE_BYTES; size > 0; size -= regSize, srcOffset += regSize, dstOffset += regSize)
     {
-        if ((size & 8) != 0)
+        while (regSize > size)
         {
-            genCodeForLoadOffset(INS_ldr, EA_8BYTE, tmpReg, srcAddr, offset);
-            genCodeForStoreOffset(INS_str, EA_8BYTE, tmpReg, dstAddr, offset);
-            offset += 8;
+            regSize /= 2;
         }
-        if ((size & 4) != 0)
+
+        instruction loadIns;
+        instruction storeIns;
+        emitAttr    attr;
+
+        switch (regSize)
         {
-            genCodeForLoadOffset(INS_ldr, EA_4BYTE, tmpReg, srcAddr, offset);
-            genCodeForStoreOffset(INS_str, EA_4BYTE, tmpReg, dstAddr, offset);
-            offset += 4;
+            case 1:
+                loadIns  = INS_ldrb;
+                storeIns = INS_strb;
+#ifdef _TARGET_ARM64_
```

Comment:
FWIW this is un unfortunate inconsistency between ARM32 and ARM64. The ARM64 emitter gets it wrong because the instruction/register size is really 4 or 8, never 1 or 2. I fixed something like a while ago for cast instructions but I didn't realize that load/store instructions suffer from the same issue.